sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala

```diff
@@ -348,13 +349,23 @@ object CreateDataSourceTableUtils extends Logging {
         className = provider,
         options = options)
 
-    def newSparkSQLSpecificMetastoreTable(): CatalogTable = {
+    def newSparkSQLSpecificMetastoreTable(relation: HadoopFsRelation): CatalogTable = {
       CatalogTable(
         identifier = tableIdent,
         tableType = tableType,
         schema = Nil,
         storage = CatalogStorageFormat(
-          locationUri = None,
+          // We don't want Hive metastore to implicitly create a table directory,
+          // which may be not the one Data Source table is referring to,
+          // yet which will be left behind when the table is dropped for an external table
+          locationUri = if (isExternal) {
+            Some(relation.location.paths.map(_.toUri.toString).map {
+              case p if new File(p).isDirectory => p
+              case p => new File(p).getParent
```

Comment:
@liancheng In this code path, we are creating a non-hive compatible metastore table, which means we don't expect users to query the table from Hive directly and get the correct results, right? And the purpose of populating this locationURI is to avoid hive metastore to implicitly create a directory.  A query against the table from spark sql will use the path set in `CatalogTable.storage.serdeProperties.get("path")`. So I think it is OK that the single file lives in a directory containing multiple files, right?